src/register/pmpaddrx.rs

```diff
@@ -27,30 +27,51 @@ impl PmpAddr {
     }
 
     #[inline]
-    pub fn encode(&mut self, mode: Mode, addr: Addr, size: Option<NonZeroSize>) {
+    pub fn encode(&mut self, mode: Mode, addr: Addr, size: Option<NonZeroSize>) -> Result<(), ()> {
         self.bits = match mode {
             Mode::OFF => 0,
-            Mode::TOR => (addr >> 2).try_into().unwrap(),
-            Mode::NA4 => (addr >> 2).try_into().unwrap(),
-            Mode::NAPOT => Self::encode_napot(addr, size.unwrap().into()),
-        }
+            Mode::TOR => {
+                let addr_small: usize = (addr >> 2) as usize;
+                if (addr_small as u64) << 2 != addr {
+                    return Err(());
+                }
+                (addr >> 2).try_into().unwrap()
+            }
+            Mode::NA4 => {
+                let addr_small: usize = (addr >> 2) as usize;
+                if (addr_small as u64) << 2 != addr {
+                    return Err(());
+                }
+                (addr >> 2).try_into().unwrap()
+            }
+            Mode::NAPOT => Self::encode_napot(addr, size.unwrap().into())?,
+        };
+        Ok(())
     }
 
     #[inline]
-    fn encode_napot(addr: Addr, size: Size) -> usize {
+    fn encode_napot(addr: Addr, size: Size) -> Result<usize, ()> {
+        // verify size is not too big
+        if (((size - 1) >> 3) > usize::MAX as Size) ||
```

Comment:
nit: The sequence `(size - 1) >> 3` is repeated quite a bit in this function. Probably should add a variable for it?

Reply:
added `encoded_size`